### PR TITLE
MPS Implementation p2. Single qubit gates

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Adds `mps_simulator.h` and single qubit gate functionality + tests. Next todo will be two qubit gates.

### CODE REVIEW

**Comment by 95-martin-orion** [comment]
@MichaelBroughton could you take a look at the failing test? It seems that Docker can't find the `eigen` files for inclusion.

**Comment by MichaelBroughton** [comment]
PTAL

File: lib/mps_simulator.h [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+#include "../eigen/Eigen/Dense"
+#include "mps_statespace.h"
+
+namespace qsim {
+
+namespace mps {
+
+/**
+ * Quantum circuit simulator without vectorization.
```
Comment (sergeisakov) [inline lib/mps_simulator.h:35]: Add "Matrix Product State (MPS)" here.

File: lib/mps_simulator.h [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+ */
+template <typename For, typename fp_type = float>
+class MPSSimulator final {
+ public:
+  using MPSStateSpace_ = MPSStateSpace<For, fp_type>;
```
Comment (sergeisakov) [inline lib/mps_simulator.h:40]: You probably can just write `using MPSStateSpace = mps::MPSStateSpace<For, fp_type>;`. So `MPSStateSpace` can be without trailing underscore here.

File: lib/mps_simulator.h [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      //   break;
+      // case 6:
+      //   ApplyGate6(qs, matrix, state);
+      //   break;
+      default:
+        // Not implemented.
+        break;
+    }
+  }
+
```
Comment (sergeisakov) [inline lib/mps_simulator.h:90]: Make the following three functions private?

**Review by sergeisakov — CHANGES_REQUESTED** [review]

File: lib/mps_simulator.h (hunk 1 above)
Comment (MichaelBroughton) [reply lib/mps_simulator.h]: Done.

File: lib/mps_simulator.h (hunk 2 above)
Comment (MichaelBroughton) [reply lib/mps_simulator.h:40]: This gave me an error saying "Cannot re-alias": 
```
 error: declaration of 'using MPSStateSpace = class qsim::mps::MPSStateSpace<For, fp_type>' changes meaning of 'MPSStateSpace' [-fpermissive]
```

File: lib/mps_simulator.h (hunk 3 above)
Comment (MichaelBroughton) [reply lib/mps_simulator.h:90]: Oops. Done.

**Review by sergeisakov — APPROVED** [review]
LGTM

**Pull request merged**